### PR TITLE
implement torch dataloader

### PR DESCRIPTION
**Pull request by haeggee**
replace the simple iid sampling at each step with a torch dataloader that has its proper seed argument.
the implementation is also in conjunction with the distributed_backend: it uses the DistributedSampler that splits the dataset properly across the processes.

this allows not only reproducibility, but also fixes different models to all see the same data (in the same order) and uses sampling without replacement

### CODE REVIEW

File: src/optim/base.py [hunk 1]
```diff
@@ -28,7 +46,7 @@ def train_base(model, opt, data, scheduler, iterations, acc_steps, batch_size, s
     t0 = time.time()
     while itr < iterations:
         for microstep_idx in range(acc_steps):  # gradient accumulation
-            x, y = get_batch(data['train'], sequence_length, batch_size, device=extra_args.device)
+            x, y = get_batch(data_train_iter, device=extra_args.device)
```
Comment (mkrima) [inline src/optim/base.py:49]: I think this will fail as soon as we end one epoch as the iterator will run out of data and raise StopIteration. Similarly the problem can happen for eval.

**Review by mkrima — CHANGES_REQUESTED** [review]
Handling epochs:
- [ ] call iter again at the beginning of every epoch
- [ ] return sampler from get_dataloader and call set_epoch

**Comment by haeggee** [comment]
implemented the changes! and also realized that the loader wasn't actually splitting the dataset into chunks of length `sequence_length`, so now the idea of an "epoch" is actually meaningful :)

**Comment by mkrima** [comment]
LGTM !
